synapse/lib/lmdb.py

```diff
@@ -0,0 +1,361 @@
+        '''
+        Check for the existence of an encoded prop, valu pair in a PropStor.
+
+        Args:
+            penc (bytes): The encoded property name.
+            byts (bytes): The valu bytes.
+
+        Returns:
+            (bool): True if the pair exists, False otherwise.
+        '''
+        return self.burs.set_key(penc + b'\x00' + byts)
+
+    def set(self, buid, penc, lval, flags=0):
+
+        pkey = buid + penc
+
+        noindex = flags & STOR_FLAG_NOINDEX
+        multival = flags & STOR_FLAG_MULTIVAL
+
+        if self.purs.set_key(pkey):
+
+            # At a minimum, we have the key...
+            if flags & STOR_FLAG_DEFVAL:
+                return False
+
+            # if this is a multi-val, iter to check for valu
+            if multival:
+
+                # if a multival prop=valu exists, skip adding
+                for mval in self.purs.iternext_dup():
+                    if mval == lval:
+                        return False
+
+            else:
+
+                # if it's exactly what we want, skip...
+                oldb = self.purs.value()
+                if oldb == lval:
+                    return False
+
+                okey = penc + b'\x00' + oldb
+                if self.burs.set_key(okey):
+                    self.burs.delete()
+
+        # we are now free and clear to set and index
+        self.purs.put(pkey, lval, dupdata=multival)
+        if not noindex:
+            self.burs.put(penc + b'\x00' + lval, buid)
+
+        return True
+
+    def put(self, items):
+        '''
+        Put a list of (buid, ((penc, lval, flags),...)) tuples.
+
+        Yields:
+            (buid, <props>) edits (only yields changes).
+        '''
+        for buid, props in items:
+
+            edits = []
+
+            for penc, lval, flags in props:
+
+                if self.set(buid, penc, lval, flags=flags):
+                    edits.append((penc, lval, flags))
+
+            if edits:
+                yield (0, (buid, edits))
+
+class PropStor:
+
+    def __init__(self, lenv, name=b'stor', edits=False):
+
+        self.lenv = lenv
+        self.edits = None
+
+        #self.tags = self.lenv.open_db(b'tags')                             # <tag>00<form>=<init><tick><tock>
+        self.props = self.lenv.open_db(name + b':props', dupsort=True)      # <buid><pkey>=<pval>
+        self.byprop = self.lenv.open_db(name + b':byprop', dupsort=True)    # <pkey>00<pval> = <buid>
+
+    def getPropSetr(self, xact):
+        return PropSetr(self, xact)
+
+    def has(self, xact, penc, byts):
+        with xact.cursor(db=self.byprop) as burs:
+            return burs.set_key(penc + b'\x00' + byts)
+
+    def pref(self, xact, penc, byts):
+        '''
+        Perform a prefix search and yield (buid, penc, pval) rows.
+        '''
+        bkey = penc + b'\x00' + byts
+
+        blen = len(bkey)
+        with xact.cursor(db=self.byprop) as burs:
+
+            if not burs.set_range(bkey):
```

Comment:
`Cursor`'s `set_range` seeks to the first key greater than or equal to the key, and will only return False if the key was past the end of the database.

Reply:
I tried sending in a bkey here that does not exist, but it still returns True. Confirmed that self.has(... returns False for it. Hmm...

Reply:
I logged the `bkey` and `burs.item()` right after the call to `burs.set_range(bkey)` returned True. The item I get back does not contain the bkey and is not the correct data. :man_shrugging: